### PR TITLE
oo so swifty

### PR DESCRIPTION
**Pull request by hisaac**
Just for fun, I tried my hand at "Swiftifying" this a bit. I don't expect you to merge it, just figured I'd show you the experiment. It's arguable whether or not this is even "better" — it's definitely more complex. Maybe safer? 🤷‍♂️

### CODE REVIEW

File: Sources/bitrise-step-swift-clone/Config.swift [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import Foundation
+
+struct Config {
+    let repositoryURL: String
+    let cloneDestination: String
+    let checkoutBranch: String?
```
Comment (russellbstephens) [inline Sources/bitrise-step-swift-clone/Config.swift:6]: I wonder if these should be encapsulated into a inner struct called Options for checkoutBranch and clone depth.  Also maybe there is a real life example to draw from

File: Sources/bitrise-step-swift-clone/Config.swift [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        self.repositoryURL = repositoryURL
+        self.cloneDestination = cloneDestination
+        self.checkoutBranch = environment[EnvKey.checkoutBranch.rawValue]
+        self.cloneDepth = environment[EnvKey.cloneDepth.rawValue]
+    }
+}
+
+enum ConfigParsingError: Error {
```
Comment (russellbstephens) [inline Sources/bitrise-step-swift-clone/Config.swift]: should we namespace this to the Config struct?

File: Sources/bitrise-step-swift-clone/Step.swift [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        if let checkoutBranch = config.checkoutBranch {
+            args.append(contentsOf: ["--branch", checkoutBranch])
+        }
+        if let cloneDepth = config.cloneDepth {
+            args.append(contentsOf: ["--depth", cloneDepth])
+        }
+        return args
+    }
+
+    func createClonePath() throws -> AbsolutePath {
```
Comment (russellbstephens) [inline Sources/bitrise-step-swift-clone/Step.swift:19]: is it more swifty to do -> var createClonePath: Result<AbsolutePath, Error>

File: Sources/bitrise-step-swift-clone/Step.swift [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        let currentPath = AbsolutePath(FileManager.default.currentDirectoryPath)
+        let clonePath = currentPath.appending(component: config.cloneDestination)
+        try FileManager.default.createDirectory(atPath: clonePath.pathString, withIntermediateDirectories: true)
+        return clonePath
+    }
+
+    func cloneRepo(into destination: AbsolutePath) throws {
```
Comment (russellbstephens) [inline Sources/bitrise-step-swift-clone/Step.swift:26]: should this live on the step? or on a FileSystem or Git struct?

File: Sources/bitrise-step-swift-clone/main.swift [hunk 5]
```diff
@@ -0,0 +1,23 @@
+import Foundation
+
+public struct bitrise_step_swift_clone {
+    public static func main() {
+        do {
```
Comment (russellbstephens) [inline Sources/bitrise-step-swift-clone/main.swift]: maybe pull this out into a `Runner` class or something to inject in the try closure

**Comment by matrangam** [comment]
This is great work! Thanks for doing this.
Once you and Russell address the issues I'm happy to bring this in.

File: Sources/bitrise-step-swift-clone/Config.swift (hunk 2 above)
Comment (hisaac) [reply Sources/bitrise-step-swift-clone/Config.swift]: Good idea, will do.

File: Sources/bitrise-step-swift-clone/Step.swift (hunk 3 above)
Comment (hisaac) [reply Sources/bitrise-step-swift-clone/Step.swift:19]: Maybe, yeah, I can try that. I honestly haven't sued the `Result` type all that much, but I think it is more "swifty" as you say.

Comment (hisaac) [reply Sources/bitrise-step-swift-clone/Step.swift:19]: I tried implementing things this way, but didn't really like the results. Lots of extra branching that makes the code more confusing imo.

I think using that `Result` type makes sense when you want to handle the `.failure()` case in a specific way, but in this case, we just want to bubble up and throw the error. So to me, just throwing it is a better choice. What do you think?

File: Sources/bitrise-step-swift-clone/Step.swift (hunk 4 above)
Comment (hisaac) [reply Sources/bitrise-step-swift-clone/Step.swift:26]: Open to discussion here, but I think that would make sense if the Step were more complex — i.e. if it were doing more Git operations than just cloning. As it is though, I think it's fine where it is. What do you think?

File: Sources/bitrise-step-swift-clone/Config.swift (hunk 2 above)
Comment (hisaac) [reply Sources/bitrise-step-swift-clone/Config.swift]: Implemented here: https://github.com/matrangam/bitrise-step-swift-clone/pull/3/commits/3603ef41f5c3f17df80b67538f170436fd51c319

File: Sources/bitrise-step-swift-clone/main.swift (hunk 5 above)
Comment (hisaac) [reply Sources/bitrise-step-swift-clone/main.swift]: Is this what you were thinking? https://github.com/matrangam/bitrise-step-swift-clone/pull/3/commits/de2c9aead13442992de88977efdd34681e8d72ac

**Review by russellbstephens — APPROVED** [review]

**Pull request merged**